### PR TITLE
Deprecate legacy configuration usages

### PR DESCRIPTION
**Pull request by jjohannes**
This PR implements deprecations around _configurations_ for `6.0` and fixes all usages of these deprecated features in the Gradle code base.
Some tests require adjustments to accept the deprecation warnings.

Each deprecation/fix is encapsulated in one commit. While the commits depend on each other regarding functionality, they can be reviewed independently.

### Goal

The goal of this is to take a step towards a better modeling of the concepts that are now covered by _configuration_. Currently, a configuration can be used for three different things and we want to reach a clearer separation, which is required to move fully to _variant-aware dependency management_. The "things" are:

1. *bucket of dependencies* (canBeResolved=false, canBeConsumed=false)
This is to declare dependencies, e.g. `implementation` or `api`
2. *export for consumption* (canBeResolved=false, canBeConsumed=true)
This is to export dependencies transitively to other projects/modules, e.g. `apiElements` or `runtimeElements` 
3. *resolve to get artifacts* (canBeResolved=true, canBeConsumed=false)
This is to actually collect all the artifacts to do something with them (like building a compile classpath), e.g. `compileClasspath` or `runtimeClasspath`

At the moment there are also:
4. canBeResolved=true+canBeConsumed=true - which is the traditional default
5. Case 2. or 3. but the configuration is used for declaration (case 1.) as well

At the moment we leave the general case 4 as it is, but deprecate all of its usages in the Java plugins (`compile`, `runtime`).
5 is deprecated in general.

### Deprecations & deprecation mechanism

- 8c1c8ff1fd9a94c535588e3007f7da6c1410d22e: Deprecate fields/methods that give access to legacy configuration names, complementing what was started in fd465dd
- 3a04ba8438a94ac2f6f4be0a8d99384f204acaae: Add deprecation mechanism for configurations, allowing configurations to be deprecated for _dependency declaration_, _consumption_, or _resolution_ independently
- Part of ☝️ 3a04ba8438a94ac2f6f4be0a8d99384f204acaae: Deprecate declaration of dependencies of _consume-only_ or _resolve-only_ configurations.
- 74d572a6184e6fad052df66637b74e6fb12b5833: Deprecation logging messages for deprecated configurations; these also end up in build scans.
- 6dcbc10e784d4edd1f68f57263d965c20e964ac9: Actually emit the messages ☝️ when a deprecation has been done
- f311fcd1634aafe9376feba356297adb0426c5cc: Finally deprecate everything legacy in the Java plugins' configurations 🎉 

### Remove usage of deprecated features

- b50e61f0d5f30d3e3f5dd4442a8cc9ec563ca326: Use 'java-library' and 'api' as defaults for buildSrc, replacing the usage of `compile`
- 75f633c35e8d9945265cfeacd8d1607f402cd9d3: Use 'java-library' and 'api' in JavaGradlePluginPlugin, replacing the usage of `compile` (this addresses #1494)
- b9d8a1d00b5014f60e644e42e9020815d39a9cc1: Adjust documentation about 'compileOnly' , hiding the old functionalities of it that are now deprecated
- 41057d2a8f0594b351efb225528f1ea1840e2904: Use implementation/runtimeOnly/testImplementation in mvn init, replacing `compile`
- cdeacd393841531eb589e075bbe783f0b2a9edb4: Exclude deprecated configurations from reports; report them already the way they will be when everything that is deprecated has been removed
- fa504d56d5ad7d279a54ec2522b73cdcafe63e79: Introduce 'annotationProcessorClasspath' configuration, to use `annotationProcessor` only for declaring (and not resolving)
- ~cfa3acb2d2b3e59c75a0a73058234a4998886846: Introduce separate configurations for path resolution of native binaries (this should only be an internal change not affecting users)~
- d7b90ee92a3ce5c38ee5f1cfd4efb0ff95c9f576: Currently the `kotlin-gradle-plugin` uses a configruation for declaration and resolving. Ignore these warnings for now, as we use that plugin in a number of tests.

### Example of deprecation warnings in a build scan

This is how it looks like for the `gradle/gradle` build before #9509 (5.5 release branch).
https://e.grdev.net/s/2hqta5zifuige/deprecations
scan

### CODE REVIEW

File: subprojects/language-native/src/main/java/org/gradle/language/cpp/internal/DefaultCppBinary.java [hunk 1]
```diff
@@ -66,24 +68,32 @@ public DefaultCppBinary(Names names, ObjectFactory objects, Provider<String> bas
...
+        includePathPathConfiguration.getAttributes().attribute(DEBUGGABLE_ATTRIBUTE, identity.isDebuggable());
+        includePathPathConfiguration.getAttributes().attribute(OPTIMIZED_ATTRIBUTE, identity.isOptimized());
+        includePathPathConfiguration.getAttributes().attribute(OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE, identity.getTargetMachine().getOperatingSystemFamily());
+        includePathPathConfiguration.getAttributes().attribute(MachineArchitecture.ARCHITECTURE_ATTRIBUTE, identity.getTargetMachine().getArchitecture());
+        includePathPathConfiguration.extendsFrom(includePathConfiguration, getImplementationDependencies());
+
+        linkLibraries = configurations.create(names.withPrefix("nativeLink"));
+        linkLibraries.setVisible(false);
+        linkLibraries.setCanBeConsumed(false);
+        linkLibraries.setCanBeResolved(false);
```
Comment (lacasseio) [inline subprojects/language-native/src/main/java/org/gradle/language/cpp/internal/DefaultCppBinary.java]: ❌  Why are any of these changes required? I thought we were already using the "correct" way of dealing with variant-aware dependencies. It seems we are using a lot more layers here. The issue seems that we are using the `nativeLink` and `cppCompile` for consumption and bucket of dependencies (case 5).

I don't think the new bucket name (with `Path` suffix) make much sense for native. It should be aligned with the JVM domain. The name would make more sense as `compileOnly`, `linkOnly` and `runtimeOnly`. However, we don't have tests for those buckets and is not even a thing in native yet. It was just coincidental that it worked. I would lean toward leaving it as it is, showing a message deprecating message and working toward a better solution for pre-built binaries.

My understanding is if we leave it as is, there will be deprecation warning issued. I think it's fine for now, we can revisit those as a separate issue.

Comment (jjohannes) [reply subprojects/language-native/src/main/java/org/gradle/language/cpp/internal/DefaultCppBinary.java]: Thanks @lacasseio for taking the time. Your understanding is correct and what you suggest makes sense to me. I'll remove the change from this PR and open a separate issue.

Comment (jjohannes) [reply subprojects/language-native/src/main/java/org/gradle/language/cpp/internal/DefaultCppBinary.java]: FYI. These are the tests that trigger a deprecation warning without the change: https://builds.gradle.org/viewLog.html?buildId=23768726

**Comment by jjohannes** [comment]
We will pick this back up once we have `master` on `6.0`.

We decided hat we will most likely not do the general [`isNotMeantForDeclaration ()`](https://github.com/gradle/gradle/commit/3a04ba8438a94ac2f6f4be0a8d99384f204acaae#diff-e02f6fd142bb98653062dbd29651e4dbR1323) deprecation just yet, before we have a better user facing API/DSL (see #9834). So this will most likely be removed from this PR. Then d7b90ee (kotlin-gradle-plugin issue) is also not needed.

**Comment by jjohannes** [comment]
To be done in follow up PRs:
- generate `@Deprecated(replaceWith = ..)`annotations for DSL accessors in Kotlin DSL
- do not lock deprecated configurations if locking all configurations of a java(-library) project (removes overhead).

**Comment by jjohannes** [comment]
This PR is now up-to-date with `master` again. The only change that remains to be done is changing all mentions of `5.7` to `6.0` once that version change has been made on `master`.

The change originally introduced an additional configuration `annotationProcessorClasspath` for symmetry. However, this causes measurable performance regressions (https://builds.gradle.org/viewLog.html?buildId=25894814). Since the change is not necessarily needed, because we also [do not do this general deprecation yet](https://github.com/gradle/gradle/pull/9353/commits/58df445643025eac355e7f36408540a0508b2078), I would leave it out. I will leave the original commit and the revert in the history, so we might resurrect some of this later once we go ahead with #9834 and improvements that follow from there.

**Pull request merged**